_posts/2022-11-23-#26152.md

```diff
@@ -7,17 +7,165 @@ authors: [Xekyo]
+[#9645](https://github.com/bitcoin/bitcoin/issues/9645),
+[#9864](https://github.com/bitcoin/bitcoin/issues/9864), and
+[#15553](https://github.com/bitcoin/bitcoin/issues/15553).
+
+- Goal: when funding a transaction using unconfirmed inputs, in addition to funding the payment(s)
+  and this transaction's fee, also include the fees necessary to bump those unconfirmed transactions
+to the target feerate.
+
+- One naive solution would be to select all the inputs, and then calculate the cost of bumping
+  unconfirmed ancestors and add that to the fees. However, if the selected inputs cannot cover the
+cost of ancestor fees, the selection process needs to be repeated; this strategy would (re)introduce
+a looped coin selection algorithm and make changeless solutions much less common.
+
+- Consider another solution which updates the "effective value" of a coin, reducing it by the fee
+  necessary to bump its ancestor set to the target feerate: `effective_value = nValue -
+cost_to_spend - min(0, target_feerate * ancestor_size - ancestor_fees)`. For example, if the target
+feerate is 20sat/vB, the UTXO has type P2WPKH and value 100ksat, and transaction's ancestor fee and
+vsize are 1200sat and 600vB respectively, the coin's effective value would be 100000sat -
+(20sat/vB * 68vB) - (20sat/vB * 600vB - 1200sat) = 87840sat.
+
+- However, this problem is much more complicated than simply using the ancestor size and fees of
+  unconfirmed UTXOs:
+
+    - Multiple UTXOs may come from the same transaction or share the same ancestry. We don't need to
+      bump those transactions more than once.
+
+    - An ancestor's feerate may be higher than the target feerate.
+
+    - An ancestor may have a low feerate, but be already fee-bumped by another transaction in the
+      mempool.
+
+    - If the transaction replaces existing transactions in the mempool, one or multiple transactions
+      may be evicted. After these evictions, some transactions may now need bumping, and others may
+      no longer need bumping. For example, if we may be replacing a previous CPFP child with an
+      even higher feerate child.
+
+- [PR #26152](https://github.com/bitcoin/bitcoin/pull/26152) adds `MiniMiner`, a "mini" version of
+  the block assembly algorithm designed to calculate the cost to bump an unconfirmed UTXO to a
+target feerate.  It uses `MiniMiner` to include the fees to bump any unconfirmed ancestors when
+funding a transaction.
+
+- `MiniMiner` operates on a limited set of mempool entries rather than the entire mempool, and does not care
+about consensus rules such as block size and sigop limits. The limited set includes the "clusters"
+of each unconfirmed transaction from which the wallet might spend a UTXO from. A *cluster* includes
+all mempool entries "connected" to a transaction. This includes parents and children,
+ancestors and descendants, as well as "siblings" (a transaction that shares a parent) and
+"coparents," (a transaction that shares a child), etc.
```

Comment:
It might help to point out that any ancestor and descendant of any transaction in the cluster is also in the cluster:

```suggestion
all mempool entries "connected" to a transaction. This includes direct parents and children,
ancestors and descendants, as well as transitively the ancestors and descendants of any connected transaction, e.g. "siblings" (a transaction that shares a parent), "coparents," (a transaction that shares a child), etc.
```